compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,818 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core._
+import Contexts._
+import Symbols._
+import Types._
+import StdNames._
+
+import ast.tpd._
+import util.EqHashMap
+import config.Printers.init as printer
+import reporting.trace as log
+
+import Errors._
+
+import scala.collection.mutable
+
+class Objects {
+  import Semantic._
+
+// ----- Domain definitions --------------------------------
+
+  /** Abstract values
+   *
+   *  Value = TypeAbs | ObjectRef | ClassAbs | Fun | RefSet
+   *
+   *  RefSet represents a set of values which may contain values other than
+   *  RefSet. The following ordering applies for RefSet:
+   *
+   *         R_a ⊑ R_b if R_a ⊆ R_b
+   *
+   *         V ⊑ R if V ∈ R
+   *
+   */
+  sealed abstract class Value {
+    def show: String = this.toString()
+  }
+
+  /** An object abstract by its type */
+  case class TypeAbs(tp: Type) extends Value
+
+  sealed abstract class Addr extends Value {
+    def klass: ClassSymbol
+  }
+
+  /** A reference to a static object */
+  case class ObjectRef(klass: ClassSymbol) extends Addr
+
+  /** An ClassAbs of class */
```

Comment:
From the comments here, the meanings of TypeAbs and ClassAbs are not clear, and it's not clear what are the differences between TypeAbs, ObjectRef, and ClassAbs. All that's clear is that they are all somehow related to an object.

Also, the abstraction ordering is not documented (although maybe these values are all unrelated by the ordering, except for RefSet in the obvious way).

Reply:
One source of possible confusion is that the word "object" is overloaded to mean many different things:
- a runtime instance of some class
- a declaration in the program using the `object` keyword
- probably other meanings as well
It would be useful to precisely define distinct terms for these different meanings.

Reply:
Indeed the documentation and naming are confusing. I try to clarify them below:

- `ClassAbs`: abstract a value by its class (outer-sensitive, but not parameter-sensitive, similar to `Warm`)
- `TypeAbs`: abstract a value by its type (primarily used for parameters)
- `ObjectRef`: reference to a global object

When `ClassAbs` gets too complex, we widen to `TypeAbs` to ensure termination.